### PR TITLE
Add filtering for policy table parameters

### PR DESCRIPTION
**Pull request by Jack-Byrne**
Update PR to filter out remaining policy table types:

- HMILevel
- AppHMIType
- Priority
- RequestType
- RequestSubType
- moduleType

### CODE REVIEW

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 1]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+  policy_table::RequestTypes valid_request_types;
+  policy_table::RequestTypes request_types = *(app_params.RequestType);
+  for (const auto& request_type : request_types) {
+    if (request_type.is_valid()) {
+      valid_request_types.push_back(request_type);
+    }
+  }
+  if (valid_request_types.empty() && !request_types.empty()) {
+    // An empty RequestType array will allow all request types. No valid
+    // parameters are in the filtered array, so assign an uninitizlied value to
```
Comment (jacobkeeler) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: Typo in `uninitialized`

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 2]
```diff
@@ -288,10 +288,19 @@ void FilterInvalidFunctions(policy_table::Rpc& rpcs) {
...
 /**
- * @brief FilterInvalidParameters filter parameters that not present in schema
+ * @brief FilterInvalidRPCParameters filter parameters that not present in
+ * schema
  * @param rpc_parameters parameters to filter
  */
-void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
+void FilterInvalidRPCParameters(policy_table::RpcParameters& rpc_parameters) {
+  policy_table::HmiLevels valid_hmi_levels;
+  for (auto& hmi_level : rpc_parameters.hmi_levels) {
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: @JackLivio `const auto&`

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 3]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+      valid_notifications.insert(notification);
+    }
+  }
+  notifications.swap(valid_notifications);
+}
+
+/**
+ * @brief FilterInvalidApplicationParameters filter app params that are not
+ * present in schema
+ * @param object of app policy params to filter
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: @JackLivio don't forget to add parameter name as the second argument of doxygen comment. Like:
```
@param app_params object of app policy params to filter
```

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 4]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+    if (request_type.is_valid()) {
+      valid_request_types.push_back(request_type);
+    }
+  }
+  if (valid_request_types.empty() && !request_types.empty()) {
+    // An empty RequestType array will allow all request types. No valid
+    // parameters are in the filtered array, so assign an uninitizlied value to
+    // for array to be "omitted"
+    policy_table::RequestTypes omitted_request_types;
+    *(app_params.RequestType) = omitted_request_types;
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: @JackLivio could we write just:
```
*(app_params.RequestType) = policy_table::RequestTypes();
```

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 5]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+  } else {
+    app_params.RequestType->swap(valid_request_types);
+  }
+
+  // Filter RquestSubTypes array
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: @JackLivio typo

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 6]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+  for (const auto& app_hmi_type : app_hmi_types) {
+    if (app_hmi_type.is_valid()) {
+      valid_app_hmi_types.push_back(app_hmi_type);
+    }
+  }
+  app_params.AppHMIType->swap(valid_app_hmi_types);
+
+  // Filter RquestTypes array
+  policy_table::RequestTypes valid_request_types;
+  policy_table::RequestTypes request_types = *(app_params.RequestType);
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: @JackLivio I think you could use reference here and add const

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 7]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+    // for array to be "omitted"
+    policy_table::RequestTypes omitted_request_types;
+    *(app_params.RequestType) = omitted_request_types;
+  } else {
+    app_params.RequestType->swap(valid_request_types);
+  }
+
+  // Filter RquestSubTypes array
+  policy_table::RequestSubTypes valid_request_sub_types;
+  policy_table::RequestSubTypes request_sub_types =
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: @JackLivio I think you could use reference here and add const

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 8]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+    if (request_sub_type.is_valid()) {
+      valid_request_sub_types.push_back(request_sub_type);
+    }
+  }
+  if (valid_request_types.empty() && !request_types.empty()) {
+    // An empty RequestSubType array will allow all request types. No valid
+    // parameters are in the filtered array, so assign an uninitizlied value to
+    // for array to be "omitted"
+    policy_table::RequestTypes omitted_request_types;
+    *(app_params.RequestType) = omitted_request_types;
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: @JackLivio see comment above regarding this

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 9]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+  } else {
+    app_params.RequestType->swap(valid_request_types);
+  }
+
+  // Filter moduleType array
+  policy_table::ModuleTypes valid_module_types;
+  policy_table::ModuleTypes module_types = *(app_params.moduleType);
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: @JackLivio see comment above regarding this

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 10]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+    app_params.moduleType->swap(valid_module_types);
+  }
+}
+
 /**
  * @brief FilterPolicyTable filter values that not present in schema
  * @param pt policy table to filter
  */
 void FilterPolicyTable(policy_table::PolicyTable& pt) {
+  policy_table::ModuleConfig& module_config = pt.module_config;
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc:398]: @JackLivio could be const?

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 11]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+  if (module_config.is_initialized() &&
+      module_config.notifications_per_minute_by_priority.is_initialized()) {
+    FilterInvalidNotifications(
+        module_config.notifications_per_minute_by_priority);
+  }
+
+  if (pt.app_policies_section.is_initialized()) {
+    policy_table::ApplicationPolicies& apps = pt.app_policies_section.apps;
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc:406]: @JackLivio could be const?

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 12]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+    for (auto& app_policy : apps) {
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc:407]: @JackLivio const auto&

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 13]
```diff
@@ -160,10 +160,19 @@ void FilterInvalidFunctions(policy_table::Rpc& rpcs) {
...
 /**
- * @brief FilterInvalidParameters filter parameters that not present in schema
+ * @brief FilterInvalidRPCParameters filter parameters that not present in
+ * schema
  * @param rpc_parameters parameters to filter
  */
-void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
+void FilterInvalidRPCParameters(policy_table::RpcParameters& rpc_parameters) {
+  policy_table::HmiLevels valid_hmi_levels;
+  for (auto& hmi_level : rpc_parameters.hmi_levels) {
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc]: @JackLivio const auto&

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 14]
```diff
@@ -173,17 +182,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+/**
+ * @brief FilterInvalidApplicationParameters filter app params that are not
+ * present in schema
+ * @param object of app policy params to filter
+ */
+void FilterInvalidApplicationParameters(
+    policy_table::ApplicationParams& app_params) {
+  // Filter AppHMIType array
+  policy_table::AppHMITypes valid_app_hmi_types;
+  policy_table::AppHMITypes app_hmi_types = *(app_params.AppHMIType);
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc]: @JackLivio const + reference

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 15]
```diff
@@ -173,17 +182,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+  for (const auto& app_hmi_type : app_hmi_types) {
+    if (app_hmi_type.is_valid()) {
+      valid_app_hmi_types.push_back(app_hmi_type);
+    }
+  }
+  app_params.AppHMIType->swap(valid_app_hmi_types);
+
+  // Filter RquestTypes array
+  policy_table::RequestTypes valid_request_types;
+  policy_table::RequestTypes request_types = *(app_params.RequestType);
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc]: @JackLivio const + reference

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 16]
```diff
@@ -173,17 +182,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+    if (request_type.is_valid()) {
+      valid_request_types.push_back(request_type);
+    }
+  }
+  if (valid_request_types.empty() && !request_types.empty()) {
+    // An empty RequestType array will allow all request types. No valid
+    // parameters are in the filtered array, so assign an uninitizlied value to
+    // for array to be "omitted"
+    policy_table::RequestTypes omitted_request_types;
+    *(app_params.RequestType) = omitted_request_types;
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc]: @JackLivio make the same changes as for policy_external according to my comments there

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 17]
```diff
@@ -173,17 +182,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+    app_params.moduleType->swap(valid_module_types);
+  }
+}
+
 /**
  * @brief FilterPolicyTable filter values that not present in schema
  * @param pt policy table to filter
  */
 void FilterPolicyTable(policy_table::PolicyTable& pt) {
+  policy_table::ModuleConfig& module_config = pt.module_config;
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc:270]: @JackLivio const

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 18]
```diff
@@ -173,17 +182,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+  if (module_config.is_initialized() &&
+      module_config.notifications_per_minute_by_priority.is_initialized()) {
+    FilterInvalidNotifications(
+        module_config.notifications_per_minute_by_priority);
+  }
+
+  if (pt.app_policies_section.is_initialized()) {
+    policy_table::ApplicationPolicies& apps = pt.app_policies_section.apps;
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc:278]: @JackLivio const

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 19]
```diff
@@ -173,17 +182,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+    for (auto& app_policy : apps) {
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc:279]: @JackLivio const auto&

**Review by AKalinich-Luxoft — CHANGES_REQUESTED** [review]

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 20]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
   rpc_parameters.parameters->swap(valid_params);
 }
 
+/**
+ * @brief FilterInvalidNotifications filter notification priorities that are not
+ * present in schema
+ * @param notification priorities to filter
+ */
+void FilterInvalidNotifications(
```
Comment (jacobkeeler) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: Name is a bit misleading. Probably should be `FirstInvalidPriorityValues` or something of the type

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 21]
```diff
@@ -301,17 +310,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+  }
+  notifications.swap(valid_notifications);
+}
+
+/**
+ * @brief FilterInvalidApplicationParameters filter app params that are not
+ * present in schema
+ * @param object of app policy params to filter
+ */
+void FilterInvalidApplicationParameters(
```
Comment (jacobkeeler) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: Might need to add logic for `priority` and `default_hmi` in this function

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 22]
```diff
@@ -173,17 +182,121 @@ void FilterInvalidParameters(policy_table::RpcParameters& rpc_parameters) {
...
+    if (request_sub_type.is_valid()) {
+      valid_request_sub_types.push_back(request_sub_type);
+    }
+  }
+  if (valid_request_types.empty() && !request_types.empty()) {
+    // An empty RequestSubType array will allow all request types. No valid
+    // parameters are in the filtered array, so assign an uninitizlied value to
+    // for array to be "omitted"
+    policy_table::RequestTypes omitted_request_types;
+    *(app_params.RequestType) = omitted_request_types;
```
Comment (jacobkeeler) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc]: This should be RequestSubType based on the context. That said, you shouldn't need to validate the RequestSubTypes, they are arbitrary strings.

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: src/components/policy/policy_external/src/policy_manager_impl.cc (hunk 10 above)
Comment (Jack-Byrne) [reply src/components/policy/policy_external/src/policy_manager_impl.cc:398]: Not this one, module_config gets modified

File: src/components/policy/policy_external/src/policy_manager_impl.cc (hunk 11 above)
Comment (Jack-Byrne) [reply src/components/policy/policy_external/src/policy_manager_impl.cc:406]: Not this one, apps gets modified

File: src/components/policy/policy_external/src/policy_manager_impl.cc (hunk 12 above)
Comment (Jack-Byrne) [reply src/components/policy/policy_external/src/policy_manager_impl.cc:407]: app_policy gets modified

File: src/components/policy/policy_regular/src/policy_manager_impl.cc (hunk 13 above)
Comment (JackLivio) [reply src/components/policy/policy_regular/src/policy_manager_impl.cc]: hmi_level gets modified

File: src/components/policy/policy_regular/src/policy_manager_impl.cc (hunk 17 above)
Comment (Jack-Byrne) [reply src/components/policy/policy_regular/src/policy_manager_impl.cc:270]: module_config gets modified

File: src/components/policy/policy_regular/src/policy_manager_impl.cc (hunk 18 above)
Comment (Jack-Byrne) [reply src/components/policy/policy_regular/src/policy_manager_impl.cc:278]: apps gets modified

File: src/components/policy/policy_regular/src/policy_manager_impl.cc (hunk 19 above)
Comment (Jack-Byrne) [reply src/components/policy/policy_regular/src/policy_manager_impl.cc:279]: app_policy gets modified

**Comment by Jack-Byrne** [comment]
@jacobkeeler @AKalinich-Luxoft Comments addressed

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 23]
```diff
@@ -288,30 +288,133 @@ void FilterInvalidFunctions(policy_table::Rpc& rpcs) {
...
+  for (const auto& hmi_level : rpc_parameters.hmi_levels) {
+    if (hmi_level.is_valid()) {
+      valid_hmi_levels.push_back(hmi_level);
+    }
+  }
+  rpc_parameters.hmi_levels.swap(valid_hmi_levels);
+
   policy_table::Parameters valid_params;
-  for (auto& param : *(rpc_parameters.parameters)) {
+  for (const auto& param : *(rpc_parameters.parameters)) {
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: @JackLivio I would suggest you to init const reference to `*(rpc_parameters.parameters)` and use it in the loop

File: src/components/policy/policy_external/src/policy_manager_impl.cc [hunk 24]
```diff
@@ -288,30 +288,133 @@ void FilterInvalidFunctions(policy_table::Rpc& rpcs) {
...
     if (param.is_valid()) {
       valid_params.push_back(param);
     }
   }
   rpc_parameters.parameters->swap(valid_params);
 }
 
+/**
+ * @brief FilterInvalidPriorityValues filter notification priorities that are
+ * not
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_external/src/policy_manager_impl.cc]: @JackLivio please merge this into the single line

File: src/components/policy/policy_regular/include/policy/policy_table/enums.h [hunk 25]
```diff
@@ -169,7 +170,8 @@ bool EnumFromJsonString(const std::string& literal, ModuleType* result);
 /**
  * @brief Enumeration FunctionID.
  *
- * Enumeration linking function names with function IDs in AppLink protocol. Assumes enumeration starts at value 0.
+ * Enumeration linking function names with function IDs in AppLink protocol.
+ *Assumes enumeration starts at value 0.
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_regular/include/policy/policy_table/enums.h]: @JackLivio please add space after `*`

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 26]
```diff
@@ -160,30 +160,133 @@ void FilterInvalidFunctions(policy_table::Rpc& rpcs) {
...
+  for (const auto& hmi_level : rpc_parameters.hmi_levels) {
+    if (hmi_level.is_valid()) {
+      valid_hmi_levels.push_back(hmi_level);
+    }
+  }
+  rpc_parameters.hmi_levels.swap(valid_hmi_levels);
+
   policy_table::Parameters valid_params;
-  for (auto& param : *(rpc_parameters.parameters)) {
+  for (const auto& param : *(rpc_parameters.parameters)) {
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc]: @JackLivio same as above regarding that

File: src/components/policy/policy_regular/src/policy_manager_impl.cc [hunk 27]
```diff
@@ -160,30 +160,133 @@ void FilterInvalidFunctions(policy_table::Rpc& rpcs) {
...
     if (param.is_valid()) {
       valid_params.push_back(param);
     }
   }
   rpc_parameters.parameters->swap(valid_params);
 }
 
+/**
+ * @brief FilterInvalidPriorityValues filter notification priorities that are
+ * not
```
Comment (AKalinich-Luxoft) [inline src/components/policy/policy_regular/src/policy_manager_impl.cc]: @JackLivio same as above regarding that

**Review by AKalinich-Luxoft — CHANGES_REQUESTED** [review]

**Comment by AKalinich-Luxoft** [comment]
@JackLivio please check for my additional comment

**Review by jacobkeeler — APPROVED** [review]

**Comment by Jack-Byrne** [comment]
@AKalinich-Luxoft Updated. Please see commit https://github.com/LuxoftAKutsan/sdl_core/pull/281/commits/3eb89f8a33c603f7d3bcee5592b62840fb728d34

**Review by AKalinich-Luxoft — APPROVED** [review]

**Comment by AKalinich-Luxoft** [comment]
@JackLivio thanks. Approved

**Comment by Jack-Byrne** [comment]
@AKalinich-Luxoft Do you have write permissions and can merge this PR?

**Comment by Jack-Byrne** [comment]
opening a pr against release/5.0.0

**Pull request closed without merging**